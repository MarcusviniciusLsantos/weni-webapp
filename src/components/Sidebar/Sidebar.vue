<template>
  <section :class="['sidebar', { 'sidebar--is-expanded': isExpanded }]">
    <RouterLink
      :to="{ name: 'orgs' }"
      class="sidebar__logo"
    >
      <section class="sidebar__logo-outer">
        <img src="@/assets/brand-name.svg" />
      </section>
    </RouterLink>

    <SidebarOption
      :option="{
        label: project.name,
        icon: 'folder',
      }"
      useDropdown
      :isExpanded="isExpanded"
      iconRight="expand_all"
      useEllipsis
      :tooltipText="$t('NAVBAR.PROJECTS')"
    >
      <template slot="dropdown-content">
        <section class="projects">
          <section class="projects__title">{{ $t('NAVBAR.PROJECTS') }}</section>

          <section class="projects__list">
            <SidebarOption
              v-for="(option, index) in projects.data"
              :key="index"
              :option="option"
              :isExpanded="true"
              isInDropdown
              useEllipsis
            />

            <template v-if="projects.status === 'loading'">
              <UnnnicSkeletonLoading
                v-for="i in 2"
                :key="i"
                tag="div"
                height="38px"
              />
            </template>
          </section>

          <footer class="projects__footer">
            <SidebarOption
              :option="{
                label: $t('NAVBAR.PROJECT_CREATE'),
                icon: 'add',
                viewUrl:
                  '/orgs/70174dd0-9ae0-4115-8ec0-951eed15047b/projects/create',
              }"
              :isExpanded="true"
              isInDropdown
              align="center"
            />

            <SidebarOption
              :option="{
                label: $t('NAVBAR.ALL_PROJECTS'),
                viewUrl: '/orgs/70174dd0-9ae0-4115-8ec0-951eed15047b/projects',
              }"
              :isExpanded="true"
              isInDropdown
              align="center"
            />
          </footer>
        </section>
      </template>
    </SidebarOption>

    <section class="pages">
      <section
        v-for="(group, index) in options"
        class="page-group"
        :key="index"
      >
        <template v-for="(option, index) in group">
          <SidebarOption
            :option="option"
            :key="index"
            :isExpanded="isExpanded"
          />
        </template>
      </section>
    </section>

    <footer class="sidebar__footer">
      <SidebarOption
        :option="{
          label: isExpanded ? $t('SIDEBAR.HIDE') : $t('SIDEBAR.SHOW'),
          icon: 'chevron_right',
        }"
        :isExpanded="isExpanded"
        variant="static"
        :iconRotate180deg="isExpanded"
        @click.native="isExpanded = !isExpanded"
      />
    </footer>
  </section>
</template>

<script>
export default {
  name: 'SidebarComponent',
};
</script>

<script setup>
import { get } from 'lodash';
import { computed, getCurrentInstance, reactive, ref, watch } from 'vue';
import SidebarOption from './SidebarOption.vue';
import gifStudio from '../../assets/tutorial/sidebar-studio.gif';
import gifIntelligences from '../../assets/tutorial/sidebar-intelligences.gif';
import gifChats from '../../assets/tutorial/sidebar-chats.gif';
import gifIntegrations from '../../assets/tutorial/sidebar-integrations.gif';
import i18n from '../../utils/plugins/i18n.js';
import APIProjects from '../../api/projects.js';
import { PROJECT_ROLE_CHATUSER } from '../users/permissionsObjects.js';

/*
  For test compatibility reasons, "store" and "route" are used as computeds.
  When possible, change this to "useStore" and "useRoute" composables.
*/

const store = computed(() => {
  const { proxy } = getCurrentInstance();
  const store = proxy.$store;
  return store;
});

const route = computed(() => {
  const { proxy } = getCurrentInstance();
  const route = proxy.$route;
  return route;
});

const props = defineProps({
  unreadMessages: Number,
});

const isExpanded = ref(true);

const projects = reactive({
  status: null,
  data: [],
});

const project = computed(() => store.value.getters.currentProject);

watch(
  () => store.value.getters.currentOrg?.uuid,
  (orgUuid) => {
    if (orgUuid) {
      loadProjects({ orgUuid });
    }
  },
  { immediate: true },
);

async function loadProjects({ orgUuid }) {
  projects.status = null;
  projects.data = [
    {
      label: project.value.name,
      viewUrl: `/projects/${get(project.value, 'uuid')}`,
      type: 'isActive',
    },
  ];

  try {
    projects.status = 'loading';

    const { data } = await APIProjects.v2List({
      params: {
        organization: orgUuid,
        offset: 0,
        limit: 6,
        ordering: '-created_at',
      },
    });

    projects.data.push(
      ...data.results
        .filter(({ uuid }) => uuid !== project.value.uuid)
        .slice(0, 5)
        .map(({ name, uuid }) => ({
          label: name,
          viewUrl: `/projects/${uuid}`,
          type: 'isActive',
        })),
    );
  } finally {
    projects.status = null;
  }
}

const isToContract = computed(() => {
  return route.value.meta?.forceContractedSidebar;
});

watch(
  () => route.value.path,
  () => {
    if (isToContract.value) {
      isExpanded.value = false;
    }
  },
  { immediate: true },
);

const hasFlows = computed(() => {
  const championChatbot =
    store.value.state.Project.championChatbots[
      store.value.getters.currentProject?.flow_organization
    ];

  return championChatbot?.error || championChatbot?.has_flows;
});

const options = computed(() => {
  const isDisabled = !hasFlows.value;

  const chatsModule = {
    label: i18n.t('SIDEBAR.chats'),
    icon: 'forum',
    viewUrl: `/projects/${get(project.value, 'uuid')}/chats`,
    type: 'isActive',
    hasNotification: !!props.unreadMessages,
    disabled: isDisabled,
    disabledModal: {
      title: i18n.t('SIDEBAR.modules.chats.title'),
      description: i18n.t('SIDEBAR.modules.chats.description'),
      image: gifChats,
    },
  };

  const settingsModule = {
    label: i18n.t('SIDEBAR.CONFIG'),
    icon: 'settings',
    viewUrl: `/projects/${get(project.value, 'uuid')}/settings`,
    type: 'isActive',
  };

  const isRoleChatUser =
    store.value.getters.currentProject.authorization.role ===
    PROJECT_ROLE_CHATUSER;

  if (isRoleChatUser) {
    return [[chatsModule], [settingsModule]];
  }

  return [
    [
      {
        label: i18n.t('SIDEBAR.HOME'),
        icon: 'home',
        viewUrl: `/projects/${get(project.value, 'uuid')}`,
        type: 'isExactActive',
      },
      {
        label: i18n.t('SIDEBAR.INSIGHTS'),
        icon: 'monitoring',
        viewUrl: `/projects/${get(project.value, 'uuid')}/insights`,
        tag: 'Beta',
        type: 'isActive',
      },
    ],
    [
      {
        label: i18n.t('SIDEBAR.BH'),
        icon: 'neurology',
        type: 'isActive',
        children: [
          {
            label: i18n.t('SIDEBAR.BRAIN'),
            viewUrl: `/projects/${get(project.value, 'uuid')}/brain`,
            tag: 'Beta',
            type: 'isActive',
          },
          {
            label: i18n.t('SIDEBAR.CLASSIFICATION_AND_CONTENT'),
            viewUrl: `/projects/${get(project.value, 'uuid')}/bothub`,
            type: 'isActive',
            disabled: isDisabled,
            disabledModal: {
              title: i18n.t('SIDEBAR.modules.intelligences.title'),
              description: i18n.t('SIDEBAR.modules.intelligences.description'),
              image: gifIntelligences,
            },
          },
        ],
      },
      {
        label: i18n.t('SIDEBAR.PUSH'),
        icon: 'account_tree',
        viewUrl: `/projects/${get(project.value, 'uuid')}/push`,
        type: 'isActive',
      },
      {
        label: i18n.t('SIDEBAR.STUDIO'),
        icon: 'ad',
        viewUrl: `/projects/${get(project.value, 'uuid')}/studio`,
        type: 'isActive',
        disabled: isDisabled,
        disabledModal: {
          title: i18n.t('SIDEBAR.modules.studio.title'),
          description: i18n.t('SIDEBAR.modules.studio.description'),
          image: gifStudio,
        },
      },
      chatsModule,
    ],
    [
      {
        label: i18n.t('SIDEBAR.INTEGRATIONS'),
        icon: 'browse',
        viewUrl: `/projects/${get(project.value, 'uuid')}/integrations`,
        type: 'isActive',
        disabled: isDisabled,
        disabledModal: {
          title: i18n.t('SIDEBAR.modules.integrations.title'),
          description: i18n.t('SIDEBAR.modules.integrations.description'),
          image: gifIntegrations,
        },
      },
      settingsModule,
    ],
  ];
});
</script>

<style lang="scss" scoped>
.pages {
  display: flex;
  flex-direction: column;
  row-gap: $unnnic-spacing-sm;
}

.page-group {
  display: flex;
  flex-direction: column;
  row-gap: $unnnic-spacing-nano;

  + .page-group {
    margin-top: -$unnnic-spacing-xs - $unnnic-border-width-thinner;
    padding-top: $unnnic-spacing-xs;
    border-top: $unnnic-border-width-thinner solid $unnnic-color-neutral-darkest;
  }
}

.sidebar {
  width: 4.5 * $unnnic-font-size;
  box-sizing: border-box;
  transition: width 100ms;

  display: flex;
  flex-direction: column;
  row-gap: $unnnic-spacing-ant;

  padding: $unnnic-spacing-sm;
  padding-top: $unnnic-spacing-ant;
  background-color: $unnnic-color-neutral-black;

  height: 100%;

  &__logo:hover {
    background-color: $unnnic-color-weni-900;
  }

  &__logo-outer {
    overflow: hidden;
    transition: height 200ms;
    height: 1.1875 * $unnnic-font-size;
  }

  &--is-expanded .sidebar__logo-outer {
    overflow: hidden;

    height: 1.25 * $unnnic-font-size;
  }

  &__logo {
    display: flex;
    align-items: center;
    height: $unnnic-icon-size-md;
    padding: $unnnic-spacing-xs;
    border-radius: $unnnic-border-radius-sm;
    user-select: none;

    img {
      height: 100%;
    }
  }

  &__footer {
    margin-top: auto;
  }

  &--is-expanded {
    width: 16.875 * $unnnic-font-size;
  }
}

.projects {
  &__title {
    color: $unnnic-color-neutral-clean;
    font-family: $unnnic-font-family-secondary;
    font-weight: $unnnic-font-weight-regular;
    font-size: $unnnic-font-size-body-md;
    line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;

    margin-bottom: $unnnic-spacing-xs;
  }

  &__list,
  &__footer {
    display: flex;
    flex-direction: column;
    row-gap: $unnnic-spacing-nano;
  }

  &__footer {
    margin-top: $unnnic-spacing-xs - $unnnic-border-width-thinner;
    border-top: $unnnic-border-width-thinner solid $unnnic-color-neutral-dark;
    padding-top: $unnnic-spacing-xs;
  }
}
</style>
