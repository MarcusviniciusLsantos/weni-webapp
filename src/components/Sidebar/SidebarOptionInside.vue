<template>
  <UnnnicToolTip
    class="tooltip"
    :enabled="enableTooltip"
    :text="tooltipText || title"
    side="right"
  >
    <component
      data-test="option-inside"
      :is="tag"
      :class="[
        'option',
        `option--align-${align}`,
        `option--variant-${variant}`,
        {
          'option--disabled': disabled,
          'option--selected': selected,
          'option--expanded': expanded,
        },
      ]"
      :href="href"
      @click="$emit('click', $event)"
    >
      <UnnnicIcon
        v-if="icon"
        :icon="icon"
        size="md"
        scheme="inherit"
        class="option__icon"
        :class="{
          'option__icon--rotate-180deg': iconRotate180deg,
        }"
        :filled="variant === 'normal' && selected"
      />

      <span
        ref="titleElement"
        class="option__title"
        :class="{
          'option__title--use-ellipsis': useEllipsis,
        }"
      >
        {{ title }}
      </span>

      <i
        v-if="label"
        class="option__tag"
      >
        {{ label }}
      </i>

      <section
        v-if="hasNotification"
        :class="[
          'option__notification',
          { 'option__notification--spaced': expanded },
        ]"
      />

      <UnnnicIcon
        v-if="iconRight"
        data-test="icon-right"
        class="option__right-icon"
        :class="{ 'option__right-icon--rotate-180deg': iconRightRotate180deg }"
        :icon="iconRight"
        size="md"
        scheme="inherit"
      />
    </component>
  </UnnnicToolTip>
</template>

<script>
export default {
  name: 'SidebarOptionInside',
};
</script>

<script setup>
import { computed, onMounted, ref } from 'vue';

const props = defineProps({
  enableTooltip: Boolean,
  tooltipText: String,
  tag: String,
  href: String,

  align: {
    type: String,
    default: 'left',
  },

  variant: {
    type: String,
    default: 'normal',
  },

  disabled: Boolean,
  selected: Boolean,
  expanded: Boolean,
  useEllipsis: Boolean,

  icon: String,
  iconRotate180deg: Boolean,

  iconRight: String,
  iconRightRotate180deg: Boolean,

  title: String,
  label: String,
  hasNotification: Boolean,

  isAccordion: Boolean,
});

const titleElement = ref();
const tooltipLeft = ref('0px');

const iconRight = computed(
  () => props.iconRight || (props.isAccordion && 'keyboard_arrow_down'),
);

onMounted(() => {
  const { offsetLeft: left, offsetWidth: width } = titleElement.value;
  const spacingXs = 8;

  tooltipLeft.value = `${left + width + spacingXs}px`;
});
</script>

<style lang="scss" scoped>
.tooltip.unnnic-tooltip {
  display: block;

  :deep(.unnnic-tooltip-label) {
    margin-left: $unnnic-spacing-sm;
  }
}

.option {
  text-decoration: none;

  display: flex;
  align-items: center;
  column-gap: $unnnic-spacing-xs;

  user-select: none;
  cursor: pointer;

  padding: $unnnic-spacing-xs;
  border-radius: $unnnic-border-radius-sm;

  white-space: nowrap;
  overflow: hidden;

  color: transparent;
  transition: color 200ms;

  font-family: $unnnic-font-family-secondary;
  font-weight: $unnnic-font-weight-regular;
  font-size: $unnnic-font-size-body-gt;
  line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;

  &--variant-outline {
    padding: $unnnic-spacing-xs - $unnnic-border-width-thinner;
    border: $unnnic-border-width-thinner solid $unnnic-color-neutral-darkest;
  }

  &--align-center {
    justify-content: center;
  }

  &--disabled {
    cursor: not-allowed;

    .option__icon,
    .option__title {
      color: $unnnic-color-neutral-dark;
    }
  }

  &__icon {
    color: $unnnic-color-neutral-clean;
    transition: transform 200ms;

    &--rotate-180deg {
      transform: rotate(-180deg);
    }
  }

  &__right-icon {
    margin-left: auto;
    color: $unnnic-color-neutral-light;
    transition: transform 200ms;

    &--rotate-180deg {
      transform: rotate(180deg);
    }
  }

  &__tag {
    margin-left: -$unnnic-spacing-nano;
    color: $unnnic-color-weni-200;
    font-weight: $unnnic-font-weight-regular;
    font-size: $unnnic-font-size-body-md;
    line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;
  }

  &__title--use-ellipsis {
    text-overflow: ellipsis;
    overflow: hidden;
  }

  &__notification {
    position: absolute;

    transition: all 200ms;

    left: 1.5625 * $unnnic-font-size;
    margin-top: -$unnnic-font-size;
    width: 0.5 * $unnnic-font-size;
    height: 0.5 * $unnnic-font-size;
    border-radius: $unnnic-border-radius-pill;
    background-color: $unnnic-color-weni-600;

    &--spaced {
      left: v-bind(tooltipLeft);
      margin-top: 0;
    }
  }
}

.option.option--expanded {
  color: $unnnic-color-neutral-light;

  &.option--variant-static {
    color: $unnnic-color-neutral-clean;
  }

  &.option--variant-normal.option--selected,
  &.option--variant-dropdown-content.option--selected {
    font-weight: $unnnic-font-weight-bold;
    color: $unnnic-color-weni-500;
  }
}

.option--variant-normal,
.option--variant-static {
  &:hover:not(.option--disabled),
  &.option--selected {
    background-color: $unnnic-color-neutral-darkest;
  }

  &.option--selected .option__icon {
    color: $unnnic-color-weni-500;
  }
}

.option--variant-dropdown-content {
  &:hover:not(.option--disabled),
  &.option--selected {
    background-color: $unnnic-color-neutral-dark;
  }
}

.option--variant-outline {
  &:hover:not(.option--disabled),
  &.option--selected {
    border: $unnnic-border-width-thinner solid $unnnic-color-neutral-clean;
  }
}
</style>
