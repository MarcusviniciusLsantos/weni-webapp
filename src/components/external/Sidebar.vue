<template>
  <div
    class="sidebar-container"
    :style="{ width: theme === 'normal' && shouldHideSideBar ? '88px' : null }"
  >
    <UnnnicSidebarPrimary
      v-if="theme === 'normal'"
      class="sidebar"
      :languages="['pt-br', 'en', 'es']"
      :language="language"
      @change-language="changeLanguage"
      :hideExpandButton="isToContract"
      :expanded="open"
      @toggle-expanded="open = $event"
      :hideText="open ? $t('SIDEBAR.HIDE') : $t('SIDEBAR.SHOW')"
      :items="categories"
      :style="
        shouldHideSideBar
          ? { position: 'absolute', top: 0, left: 0, zIndex: 2 }
          : null
      "
    >
      <template v-slot:header>
        <div class="sidebar-header">
          <RouterLink to="/orgs">
            <img src="../../assets/Logo-Weni-Soft-Default.svg" />
          </RouterLink>
        </div>
      </template>

      <template v-slot:search>
        <div v-if="open">
          <UnnnicAutocomplete
            v-if="theme == 'normal' && !hideModulesButChats"
            :placeholder="$t('NAVBAR.SEARCH_PLACEHOLDER')"
            size="sm"
            class="sidebar__search"
            iconLeft="search-1"
            v-model="search"
            :data="items"
            @input="onSearch"
            highlight
            @choose="chooseOption"
          />
        </div>
      </template>

      <template v-if="!hasFlows">
        <SidebarModal
          slot="block-studio"
          :title="$t('SIDEBAR.modules.studio.title')"
          :description="$t('SIDEBAR.modules.studio.description')"
          :image="gifStudio"
        />

        <SidebarModal
          slot="block-intelligences"
          :title="$t('SIDEBAR.modules.intelligences.title')"
          :description="$t('SIDEBAR.modules.intelligences.description')"
          :image="gifIntelligences"
        />

        <SidebarModal
          slot="block-chats"
          :title="$t('SIDEBAR.modules.chats.title')"
          :description="$t('SIDEBAR.modules.chats.description')"
          :image="gifChats"
        />

        <SidebarModal
          slot="block-integrations"
          :title="$t('SIDEBAR.modules.integrations.title')"
          :description="$t('SIDEBAR.modules.integrations.description')"
          :image="gifIntegrations"
        />
      </template>
    </UnnnicSidebarPrimary>

    <div
      v-show="theme === 'normal' && shouldHideSideBar && open"
      @click="open = false"
      class="background"
    ></div>
  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex';
import { get } from 'lodash';
import getEnv from '@/utils/env';
import ShouldShowBrain from '../../utils/ShouldShowBrain';
import { PROJECT_ROLE_CHATUSER } from '../users/permissionsObjects';
import SidebarModal from '../SidebarModal.vue';
import gifStudio from '../../assets/tutorial/sidebar-studio.gif';
import gifIntelligences from '../../assets/tutorial/sidebar-intelligences.gif';
import gifChats from '../../assets/tutorial/sidebar-chats.gif';
import gifIntegrations from '../../assets/tutorial/sidebar-integrations.gif';
import projects from '../../api/projects';

export default {
  // eslint-disable-next-line vue/multi-word-component-names
  name: 'Sidebar',

  components: {
    SidebarModal,
  },

  props: {
    unreadMessages: Number,
  },
  data() {
    return {
      items: [],
      open: true,
      current: '',
      notifyAgents: false,
      gifStudio,
      gifIntelligences,
      gifChats,
      gifIntegrations,
      innerWidth: innerWidth,
      resizeFunction: () => {
        this.innerWidth = window.innerWidth;
      },
      search: '',
      activeSearch: null,
      loading: false,
    };
  },

  created() {
    window.addEventListener('resize', this.resizeFunction);

    window.addEventListener('message', function (e) {
      if (e.data.eventName === 'unread-changed') {
        this.notifyAgents = e.data.data !== '';
      }
    });

    this.$root.$on('set-sidebar-expanded', () => {
      if (!this.isToContract) {
        this.open = true;
      }
    });
  },

  beforeDestroy() {
    window.removeEventListener('resize', this.resizeFunction);
  },

  computed: {
    ...mapGetters(['currentProject']),

    shouldHideSideBar() {
      return this.innerWidth < 768;
    },

    hasFlows() {
      const championChatbot =
        this.$store.state.Project.championChatbots[
          this.$store.getters.currentProject?.flow_organization
        ];

      return championChatbot?.error || championChatbot?.has_flows;
    },

    hideModulesButChats() {
      if (
        getEnv('MODULES_YAML').chats &&
        this.$store.getters.currentProject.authorization.role ===
          PROJECT_ROLE_CHATUSER
      ) {
        return true;
      }

      return false;
    },

    language() {
      return this.$i18n.locale;
    },

    categories() {
      const project = this.currentProject;

      const icons = {
        house: ['home'],
        hub: ['hub'],
        hierarchy: ['account_tree'],
        'app-window-edit': ['ad'],
        'layout-dashboard': ['browse'],
        'science-fiction-robot': ['neurology'],
        'messaging-we-chat': ['forum'],
        'single-neutral': ['person'],
        config: ['settings'],
        monitoring: ['monitoring'],
      };

      return [
        {
          type: 'category',
          label: 'SIDEBAR.MAIN_MENU',
          show: () => {
            return !this.hideModulesButChats;
          },
          items: [
            {
              name: 'home',
              label: 'SIDEBAR.HOME',
              icon: 'house',
              viewUrl: `/projects/${get(project, 'uuid')}`,
            },
          ],
        },
        {
          type: 'category',
          label: 'SIDEBAR.SYSTEMS',
          items: [
            {
              name: 'insights',
              label: 'SIDEBAR.INSIGHTS',
              icon: 'monitoring',
              viewUrl: `/projects/${get(project, 'uuid')}/insights/init`,
              tag: 'Alfa',
              show: () => {
                return this.$store.state.Account.profile?.email.includes(
                  '@weni.ai',
                );
              },
            },
            {
              label: 'SIDEBAR.BRAIN',
              icon: 'hub',
              viewUrl: `/projects/${get(project, 'uuid')}/brain/init`,
              show: () => {
                if (
                  !ShouldShowBrain({
                    projectUuid: get(project, 'uuid'),
                    projectCreatedAt:
                      this.$store.state.Project.currentProject?.created_at,
                    userEmail: this.$store.state.Account.profile?.email,
                  })
                ) {
                  return false;
                }

                return !this.hideModulesButChats;
              },
            },
            {
              label: 'SIDEBAR.PUSH',
              icon: 'hierarchy',
              viewUrl: `/projects/${get(project, 'uuid')}/push/init`,
              show: () => {
                return !this.hideModulesButChats;
              },
            },
            {
              label: 'SIDEBAR.STUDIO',
              id: 'studio',
              icon: 'app-window-edit',
              viewUrl: `/projects/${get(project, 'uuid')}/studio/init`,
              show: () => {
                return !this.hideModulesButChats;
              },
            },
            {
              label: 'SIDEBAR.BH',
              id: 'intelligences',
              icon: 'science-fiction-robot',
              viewUrl: `/projects/${get(project, 'uuid')}/bothub/init`,
              show: () => {
                return !this.hideModulesButChats;
              },
            },
            {
              label: 'SIDEBAR.chats',
              id: 'chats',
              icon: 'messaging-we-chat',
              viewUrl: `/projects/${get(project, 'uuid')}/chats/init`,
              show: () => {
                return getEnv('MODULES_YAML').chats;
              },
              notify: !!this.unreadMessages,
            },
          ],
        },
        {
          type: 'category',
          label: 'SIDEBAR.PROJECT',
          show: () => {
            return true;
          },
          items: [
            {
              label: 'SIDEBAR.INTEGRATIONS',
              id: 'integrations',
              icon: 'layout-dashboard',
              viewUrl: `/projects/${get(project, 'uuid')}/integrations/init`,
              show: () => {
                if (this.hideModulesButChats) {
                  return false;
                }

                return true;
              },
            },
            {
              label: 'SIDEBAR.CONFIG',
              icon: 'config',
              viewUrl: `/projects/${get(project, 'uuid')}/settings`,
              show: () => {
                return true;
              },
            },
          ],
        },
      ]
        .filter((item) => {
          if (item.show) {
            return item.show(project);
          }

          return true;
        })
        .map((item) => ({
          ...item,
          label: this.$t(item.label),
          items: item.items
            .filter((item) => {
              if (item.show) {
                return item.show(project);
              }

              return true;
            })
            .map((route) => {
              const active =
                route.name === 'home'
                  ? this.$route.name === route.name
                  : this.$route.path.startsWith(
                      route.viewUrl.replace('/init', ''),
                    );

              return {
                ...route,
                label: this.$t(route.label),
                active,
                icon: icons[route.icon][0],
                click: () => {
                  this.$router.push(route.viewUrl);
                },
                notify: route.notify,
              };
            }),
        }));
    },

    isToContract() {
      return this.$route.meta?.forceContractedSidebar;
    },
  },
  methods: {
    ...mapActions(['updateAccountLanguage']),

    changeLanguage(language) {
      this.updateAccountLanguage({ language });
    },

    pathname(context, pathSegment) {
      if (!context) return `/${pathSegment}`;
      else return `/${context}/${pathSegment}`;
    },

    onSearch() {
      if (!this.search) {
        this.items = [];
        return false;
      }

      this.loading = true;

      if (this.activeSearch) {
        clearTimeout(this.activeSearch);
      }

      const makeUrl = (type, data) => {
        const system = {
          flow: 'push',
          intelligence: 'bothub',
        };

        const base = `/projects/${this.currentProject.uuid}/${system[type]}`;

        if (type === 'flow') {
          return `${base}/f/flow/editor/${data.uuid}`;
        } else if (type === 'intelligence') {
          if (data.repository_type === 'classifier') {
            return `${base}/f/dashboard/${data.owner__nickname}/${data.slug}`;
          } else if (data.repository_type === 'content') {
            return `${base}/f/dashboard/${data.owner__nickname}/${data.slug}/content/bases`;
          }
        }
      };

      this.activeSearch = setTimeout(async () => {
        try {
          const response = await projects.search(
            null,
            this.currentProject.uuid,
            this.search,
          );

          const { data } = response;

          this.items = [];

          if (data.inteligence?.results.length) {
            this.loading = false;
            this.items.push({
              type: 'category',
              text: this.$t('SIDEBAR.BH'),
            });

            data.inteligence?.results
              .map((item) => ({
                type: 'option',
                text: item.name,
                value: {
                  ...item,
                  href: makeUrl('intelligence', item),
                },
              }))
              .forEach((item) => this.items.push(item));
          }

          if (data.flow.length) {
            this.loading = false;
            this.items.push({
              type: 'category',
              text: this.$t('SIDEBAR.PUSH'),
            });

            data.flow
              .map((item) => ({
                type: 'option',
                text: item.name,
                value: {
                  ...item,
                  href: makeUrl('flow', item),
                },
              }))
              .forEach((item) => this.items.push(item));
          }

          if (this.items.length === 0) {
            this.loading = false;
            this.items.push({
              type: 'category',
              text: this.$t('NAVBAR.NO_RESULTS'),
            });
          }
        } catch (e) {
          console.log(e);
        }
      }, 300);
    },

    chooseOption(value) {
      if (value.href) {
        this.$router.push(value.href);
      }
    },
  },

  watch: {
    '$route.path': {
      immediate: true,
      handler() {
        if (this.isToContract) {
          this.open = false;
        }
      },
    },

    shouldHideSideBar: {
      immediate: true,

      handler() {
        if (this.shouldHideSideBar && this.open) {
          this.open = false;
        }
      },
    },

    loading() {
      if (this.loading) {
        this.items = [];
        this.items.push({
          type: 'category',
          text: this.$t('NAVBAR.LOADING'),
        });
      }
    },
  },
};
</script>

<style lang="scss" scoped>
@import '~@weni/unnnic-system/src/assets/scss/unnnic.scss';
@import '~@weni/unnnic-system/dist/unnnic.css';

$transition-time: 0.4s;

.background {
  width: 100vw;
  height: 100vh;
  position: fixed;
  top: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.5);
  z-index: 1;
}

.sidebar {
  min-height: 100vh;

  &-header {
    width: 24px;
    height: $unnnic-icon-size-md;
    overflow: hidden;
    margin: 0 auto;
    transition: width $transition-time;

    img {
      vertical-align: middle;
      height: 18.89px;
      transition: all $transition-time;
    }
  }

  &__search {
    margin-bottom: $unnnic-inline-sm;
    flex: 1;
  }

  &.unnnic-sidebar-primary-expanded {
    .sidebar-header {
      width: 85px;

      img {
        height: $unnnic-icon-size-md;
      }
    }
  }
  ::v-deep .unnnic-sidebar-primary {
    padding: 32px 24px;
  }
  ::v-deep .unnnic-language-select {
    z-index: 1;
  }
}
</style>
