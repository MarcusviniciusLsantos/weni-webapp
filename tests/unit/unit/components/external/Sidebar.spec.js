import { vi } from 'vitest';
import { shallowMount, createLocalVue, RouterLinkStub } from '@vue/test-utils';
import Sidebar from '@/components/external/Sidebar.vue';
import i18n from '@/utils/plugins/i18n';

import Vuex from 'vuex';
import Router from 'vue-router';

import { project } from '../../../__mocks__';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Router);

const router = new Router();

describe('Sidebar.vue', () => {
  let wrapper;
  let getters;
  let store;
  let actions;
  let $route = {
    path: '/',
  };

  beforeEach(() => {
    getters = {
      currentProject() {
        return project;
      },
    };
    actions = {
      updateAccountLanguage: vi.fn(),
    };
    store = new Vuex.Store({ getters, actions });
    wrapper = shallowMount(Sidebar, {
      localVue,
      i18n,
      router,
      store,
      mocks: {
        $t: () => 'some specific text',
        $route,
      },
      stubs: {
        RouterLink: RouterLinkStub,
        UnnnicSidebarPrimary: true,
      },
    });
  });

  it('renders a snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('verify change language', async () => {
    await wrapper.vm.changeLanguage('es');
    expect(actions.updateAccountLanguage).toHaveBeenCalled();
  });

  it('verify pathname', async () => {
    const res = await wrapper.vm.pathname(false, 'oi');
    expect(res).toEqual('/oi');
  });
  it('verify pathname with context', async () => {
    const res = await wrapper.vm.pathname('test', 'oi');
    expect(res).toEqual('/test/oi');
  });
});
