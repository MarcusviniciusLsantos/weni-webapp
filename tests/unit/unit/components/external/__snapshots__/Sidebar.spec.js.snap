// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar.vue > renders a snapshot 1`] = `
<div
  class="sidebar-container"
  data-v-31a33fdf=""
>
  <!---->
  <div
    class="background"
    data-v-31a33fdf=""
    style="display: none;"
  />
</div>
`;
